connect_four.py

```diff
@@ -101,46 +107,102 @@ def __diagonal_pos_slope(self, row: int, col: int, color: str) -> bool:
 
+class InputOutput:
+    def __init__(self, run_mode: RunMode, moves_list: list) -> None:
+        self.run_mode = run_mode #if TRUE, user input will be asked for. Otherwise, input will be automatic from test functions
+        self.moves_list = moves_list
+        self.move_counter = 0
+
+    def player_names(self) -> str:
+        if self.run_mode is RunMode.PLAY_MODE:
+            # take player's names
+            print("Player 1, you are yellow. What is your name?")
+            player1 = str(input())
+            print("Player 2, you are red. What is your name?")
+            player2 = str(input())
+        else:
+            player1, player2 = "player1", "player2"
+
+        return player1, player2
+
+    def game_over_IO(self, winner: str, board) -> str:
+        print(str(board))
+        if self.run_mode is RunMode.PLAY_MODE:
+            if winner:
+                print(f"{winner}, you win!! Press R to restart, or press any key to exit the game.")
+                response = str(input())[0].capitalize() 
+            else: # __game_over has been called without a winner
+                print(f"The game has ended in a draw. Press R to restart, or press any key to exit the game.")
+                response = str(input())[0].capitalize()
+        else:
+            response = "Exit"
+        
+        return response
+    
+    def select_col_IO(self, board, player: str) -> int:
+        if self.run_mode is RunMode.PLAY_MODE:
+            last_col = chr(ord('A') - 1 + board.num_cols)
+
+            print(f"{player}, it is your turn. Enter the column you would like to add a piece to (A-{last_col}):")
+            col = str(input())[0].capitalize()
+
+            col_code = board.check_col(col)
+            while col_code is not ErrorCode.VALID:
+                if col_code is ErrorCode.COL_INVALID:
+                    print(f"{player}, that column does not exist. Please pick another column (A-{last_col}):") 
+                    col = str(input())[0].capitalize()
+                    col_code = board.check_col(col)
+                else:
+                    print(f"{player}, that column is full. Please pick another column (A-{last_col}):") 
+                    col = str(input())[0].capitalize()
+                    col_code = board.check_col(col)
+        
+        else:
+            col_code = None
+            while col_code is not ErrorCode.VALID:
+                col = self.moves_list[self.move_counter].capitalize()
+                self.move_counter += 1 
+                col_code = board.check_col(col)
+
+        return ord(col) - ord('A')
 
+    
 
 # Main gameplay. Set as while loop to play until exited.
 class GameLoop: 
     
-    def __init__(self) -> None:
+    def __init__(self, run_mode: RunMode = RunMode.PLAY_MODE, moves_list: list = []) -> None:
```

Comment:
Haha, there's a small issue with this, you may need to modify your code a little. Take a look here: 
https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

I chuckle because I wasn't aware of this either until recently. It's one of those classic Python little tidbits.

Reply:
After you fix this, go ahead and merge in the PR. it looks solid: Good stuff man.

Reply:
Oh this is a very interesting bug. Thanks for sharing!

Reply:
not sure if you gleaned this but It's not necessarily a bug but _can lead to a bug_

Reply:
Yeah I noticed, wasn't sure what to call it lol